[DataGridPro] Tree data row reordering support (#19401)
mui/mui-x: docs/data/data-grid/row-ordering/row-ordering.md

For now, row reordering is disabled if sorting is applied to the Data Grid.
:::

## Disable reordering of specific rows

To disable reordering of specific rows, you can pass a callback to the `isRowReorderable()` prop.
This callback is called with `row` and `rowNode` objects to allow disabling rows based on multiple criteria.

```tsx
<DataGridPro isRowReorderable={isRowReorderable} />
```

The demo below shows how to disable reordering of specific rows based on the row model.

{{"demo": "RowReorderingDisabled.js", "bg": "inline"}}

## Disable specific reorder operations

Sometimes, the information provided by the source row isn't enough to determine if a reorder operation is valid.
In these cases, you can pass a callback to the `isValidRowReorder()` prop.

This callback is fired _during_ the drag operation so it provides information about the dragged row and potential row where it is being dropped.
It is called with a `ReorderValidationContext` object to allow disabling specific reorder operations based on the context.

> The demo below prohibits reordering certain rows under the same parent and only allows cross-parent operations.

Reviewer: ```suggestion
The demo below prohibits reordering leaf rows under the same parent and only allows cross-parent operations.
```

{{"demo": "RowReorderingValidation.js", "bg": "inline"}}

:::warning
The row reorder feature has an internal validation ruleset that makes sure that only the supported use-cases are allowed.

> So the `isValidRowReorder()` callback should only be used to omit some of the supported use-cases, not adding new ones.

Reviewer: ```suggestion
`isValidRowReorder()` should only be used to omit some of the supported use-cases, not add new ones.
```

:::

:::info

> The above demo uses row grouping to demonstrate the concept. You can check more about this in the [Row grouping—Drag-and-drop group reordering](/x/react-data-grid/row-grouping/#drag-and-drop-group-reordering) documentation section.

Reviewer: Would it be better if the demo used Tree data, since it is the feature in the same plan as row reordering?

:::

## Row reordering with tree data

> This feature allows users to reorder rows within tree data to adjust the order of groups and their contents directly in the grid.

Reviewer 2: ```suggestion
Users can drag and drop rows to reorganize tree hierarchies—reorder siblings, move rows between parents, or convert leaf nodes into parent groups.
```

For now, row reordering is disabled if sorting is applied to the Data Grid.
:::

## Disable reordering of specific rows

To disable reordering of specific rows, you can pass a callback to the `isRowReorderable()` prop.
This callback is called with `row` and `rowNode` objects to allow disabling rows based on multiple criteria.

```tsx
const isRowReorderable = (params: IsRowReorderableParams) => {
  // Example: Disable reordering if the quantity is greater than 50000
  return params.row.quantity > 50000;
};

<DataGridPro isRowReorderable={isRowReorderable} />;
```

The demo below shows how to disable reordering of specific rows based on the row model.

{{"demo": "RowReorderingDisabled.js", "bg": "inline"}}

## Disable specific reorder operations

Sometimes, the information provided by the source row isn't enough to determine if a reorder operation is valid.
In these cases, you can pass a callback to the `isValidRowReorder()` prop.

> This callback is fired _during_ the drag operation so it provides information about the dragged row and potential row where it is being dropped.

Reviewer 3: IIUC I think this is a little more clear

```suggestion
This callback is fired _during_ the drag operation so it provides information about the dragged row and the potential position for it to be dropped.
```

It is called with a `ReorderValidationContext` object to allow disabling specific reorder operations based on the context.

The demo below prohibits reordering leaf rows under the same parent and only allows cross-parent operations.

{{"demo": "RowReorderingValidation.js", "bg": "inline"}}

:::warning
The row reorder feature has an internal validation ruleset that makes sure that only the supported use-cases are allowed.

> `isValidRowReorder()` should only be used to omit some of the supported use-cases, not add new ones.

Author: Don't they mean the same though 🤔
